streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -822,8 +828,51 @@ private void addStreamTasks(Collection<TopicPartition> assignment) {
             }
         }
 
-        // finally destroy any remaining suspended tasks
+        // destroy any remaining suspended tasks
         removeSuspendedTasks();
+
+        // create all newly assigned tasks (guard against race condition with other thread via backoff and retry)
+        // -> other thread will call removeSuspendedTasks(); eventually
+        long backoffTimeMs = 50L;
+        while (true) {
```

Comment:
If it was me writing this i'd probably extract this loop etc into a class. It could be a static class within `StreamThread` package-private. The main reason would be so I can easily test this code in isolation of everything else.